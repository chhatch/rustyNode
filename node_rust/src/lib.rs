extern crate wasm_bindgen;
use web_sys::console;

use wasm_bindgen::prelude::*;

mod rules;

#[wasm_bindgen]
pub fn main(data_string: String) -> String {
    console::log_1(&"Greetins from Rustlandia!".into());
    return rules::run(data_string);
}
