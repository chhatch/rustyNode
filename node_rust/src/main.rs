extern crate serde_json;
use serde::Deserialize;
use serde::Serialize;
use std::fs;
use std::io::Read;

mod operations;

#[derive(Deserialize, Debug, Serialize)]
struct Data {
    node: NODE,
    rust: RUST,
    ruby: i32,
    day_of_week: String,
    fee: i32,
    price: i32,
}
#[derive(Deserialize, Debug, Serialize)]
struct NODE {
    fp: FP,
}
type FP = Vec<bool>;
type RUST = Vec<NAIL>;
#[derive(Deserialize, Debug, Serialize)]
struct NAIL {
    nail: String,
}
fn main() {
    let mut file = fs::File::open("input.json").unwrap();
    let mut data_string = String::new();

    file.read_to_string(&mut data_string).unwrap();

    let mut parsed_data: Data =
        serde_json::from_str(&data_string).expect("JSON was not well-formatted");
    if parsed_data.node.fp[0] == true {
        parsed_data.rust[0].nail = "win".to_string()
    } else {
        parsed_data.ruby = 1337
    }
    if parsed_data.day_of_week == "Wednesday".to_string() {
        parsed_data.fee = 2
    } else {
        parsed_data.fee = 1
    }
    if parsed_data.day_of_week == "Friday".to_string() {
        parsed_data.fee = 0
    }
    if parsed_data.day_of_week != "Friday".to_string() {
        parsed_data.price = operations::pow(parsed_data.price, parsed_data.fee)
    }
    let processed_data_string = serde_json::to_string_pretty(&parsed_data).unwrap();
    println!("{}", processed_data_string);
    fs::write("output.json", processed_data_string).expect("Unable to write file");
}
