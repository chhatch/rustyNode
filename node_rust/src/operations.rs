pub fn pow(x: i32, y: i32) -> i32 {
    x.pow(y.try_into().unwrap())
}
