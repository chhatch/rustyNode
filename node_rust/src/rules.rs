extern crate serde_json;
use serde::Deserialize;
use serde::Serialize;

mod operations;

#[derive(Deserialize, Debug, Serialize)]
struct Data {
    node: NODE,
    rust: RUST,
    ruby: i32,
    day_of_week: String,
    fee: i32,
    price: i32,
}
#[derive(Deserialize, Debug, Serialize)]
struct NODE {
    fp: FP,
}
type FP = Vec<bool>;
type RUST = Vec<NAIL>;
#[derive(Deserialize, Debug, Serialize)]
struct NAIL {
    nail: String,
}
pub fn run(data_string: String) -> String {
    let mut parsed_data: Data =
        serde_json::from_str(&data_string).expect("JSON was not well-formatted");
    if parsed_data.node.fp[0] == true {
        parsed_data.rust[0].nail = "win".to_string()
    } else {
        parsed_data.ruby = 1337
    }
    if parsed_data.day_of_week == "Wednesday".to_string() {
        parsed_data.fee = 2
    } else {
        parsed_data.fee = 1
    }
    if parsed_data.day_of_week == "Friday".to_string() {
        parsed_data.fee = 0
    }
    if parsed_data.day_of_week != "Friday".to_string() {
        parsed_data.price = operations::pow(parsed_data.price, parsed_data.fee)
    }
    return serde_json::to_string_pretty(&parsed_data).unwrap();
}
