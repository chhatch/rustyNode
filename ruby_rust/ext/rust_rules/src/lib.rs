use magnus::{
    define_global_function,
    encoding::{CType, EncodingCapable, RbEncoding},
    function,
    prelude::*,
    Error, RString,
};

extern crate serde_json;
mod rules;

#[magnus::init]
fn init() -> Result<(), Error> {
    define_global_function("rules", magnus::function!(rules::run, 1));
    Ok(())
}
