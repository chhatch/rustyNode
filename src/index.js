import { readFileSync, writeFileSync } from "fs";
import { flow, partial, partialRight } from "lodash-es";
import { compileRust } from "./rustWriter/index.js";
import { parseRuleObject } from "./ruleParser/index.js";

flow([
  partialRight(readFileSync, "utf8"),
  JSON.parse,
  parseRuleObject,
  compileRust,
  partial(writeFileSync, "node_rust/src/main.rs"),
])("rule.json");
