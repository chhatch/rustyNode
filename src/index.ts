import { readFileSync, writeFileSync } from 'fs'
import { flow, partial, partialRight } from 'lodash'
import { compileRust } from './rustWriter/index'
import { parseRules } from './ruleParser/index'
import { processArgs } from './cli'
const { rustPath, rulesPath, inputPath, outputPath } = processArgs()
export const rulesToRust = flow([
  partialRight(readFileSync, 'utf8'),
  JSON.parse,
  parseRules,
  compileRust(inputPath, outputPath),
  partial(writeFileSync, rustPath)
])

rulesToRust(rulesPath)
