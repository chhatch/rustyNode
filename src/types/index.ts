export interface ParsedRule {
  [key: string]: {
    lhs: string
    operator: string
    rhs: string
    rustString: string
  }
}
export type DataTypesEnum = 'number' | 'boolean' | 'string' | 'unknown'
export interface DataStructure {
  [key: string]: {
    type: DataTypesEnum
    mutable: boolean
  }
}
export interface TermNode {
  value: string | number | boolean
  rustString: string
  type: DataTypesEnum
  lhs?: TermNode
  operator?: string
  rhs?: TermNode
}

export type NodeFlow = [TermNode, TermNode, string, TermNode]

export type UnwrappedThunks = [TermNode, string, TermNode]
