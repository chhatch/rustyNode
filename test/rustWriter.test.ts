import {
  buildRustStruct,
  compileRust,
  generateDataStructure,
} from "../src/rustWriter";
import { DataStructure, ParsedRule } from "../src/types";

describe("generateDataStructure", () => {
  it("should generate the data structure", () => {
    const rules = [
      {
        else: { lhs: "ruby", operator: "=", rhs: "1337" },
        if: { lhs: "node", operator: "==", rhs: "true" },
        then: { lhs: "rust", operator: "=", rhs: "'win'" },
      },
      {
        if: { lhs: "day_of_week", operator: "!=", rhs: "'Friday'" },
        then: { lhs: "price", operator: "=", rhs: "15" },
      },
    ] as ParsedRule[];

    const result = {
      node: { type: "boolean", mutable: false },
      rust: { type: "string", mutable: true },
      ruby: { type: "number", mutable: true },
      day_of_week: { type: "string", mutable: false },
      price: { type: "number", mutable: true },
    };

    expect(generateDataStructure(rules)).toEqual(result);
  });
});

describe("buildRustStruct", () => {
  it("should build a rust struct", () => {
    const dataStructure = {
      node: { type: "boolean", mutable: false },
      rust: { type: "string", mutable: true },
      ruby: { type: "number", mutable: true },
      day_of_week: { type: "string", mutable: false },
      price: { type: "number", mutable: true },
    } as DataStructure;

    expect(buildRustStruct(dataStructure)).toMatchInlineSnapshot(`
"#[derive(Deserialize, Debug, Serialize)]
struct Data {
node: bool,
rust: String,
ruby: i32,
day_of_week: String,
price: i32,
}
"
`);
  });
});

describe("compileRust", () => {
  it("should generate a rust file", () => {
    const rules = [
      {
        else: { lhs: "ruby", operator: "=", rhs: "1337" },
        if: { lhs: "node", operator: "==", rhs: "true" },
        then: { lhs: "rust", operator: "=", rhs: "'win'" },
      },
      {
        if: { lhs: "day_of_week", operator: "!=", rhs: "'Friday'" },
        then: { lhs: "price", operator: "=", rhs: "15" },
      },
    ] as ParsedRule[];

    expect(compileRust(rules)).toMatchSnapshot();
  });
});
