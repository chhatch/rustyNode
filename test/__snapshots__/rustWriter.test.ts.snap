// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compileRust should generate a rust file 1`] = `
"extern crate serde_json;
use serde::Deserialize;
use serde::Serialize;

mod operations;

#[derive(Deserialize, Debug, Serialize)]
struct Data {
node: bool,
rust: String,
ruby: f64,
day_of_week: String,
price: f64,

}
pub fn run(data_string: String) -> String {
    let mut parsed_data: Data =
        serde_json::from_str(&data_string).expect(\\"JSON was not well-formatted\\");
if parsed_data.node == true {
        parsed_data.rust = \\"win\\".to_string()
    }
else {
        parsed_data.ruby = 1337
    }
if parsed_data.day_of_week != \\"Friday\\".to_string() {
        parsed_data.price = 15
    }
return serde_json::to_string_pretty(&parsed_data).unwrap();
}"
`;
